<a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 環境構築メモ

**Gitインストール　https://git-scm.com/**
「Download 2.XX.x for Windows」となっているボタンをクリック<br>
※　ダウンロードが始まらなければ、手動でダウンロード<br>
ダウンロードしたインストーラをダブルクリックし、インストール<br>
・　例：Git-2.XX.x-64-bit.exe をダブルクリック<br>
<br>
Anaconda で Git のパスを通す手順<br>
Anaconda で Git を使用する場合、Anaconda 上で Git のパスを通す必要がある<br>
以下を参考<br>
①　まず、Git のパスを確認<br>
デフォルトのまま Git をインストールした場合、パスは以下になっている<br>
Windows のエクスプローラを起動して下記のパスに移動し、ファイルが入っていることを確認<br>
・　Git のパス：C:\Program Files\Git\cmd<br>
②　Anaconda Prompt を起動<br>
③　以下を参考に、Git を使いたい仮想環境をアクティベート<br>
入力例<br>
“conda info -e” → “conda activate [仮想環境名]” → [enter] キーを押す。<br>
具体例<br>
conda activate yolov5 + [enter] キーを押す。
※　あらかじめ、仮想環境 yolov5 を作っておく必要あり<br>
<br>
Anaconda Prompt で設定されているパスを確認<br>
 “path” と入力 →  [enter] キーを押す。<br>
→　設定されているパスが表示<br>
入力例
 set PATH=%PATH%;C:\Program Files\Git\cmd
<br>
以上でCondaからGitが使えるようになる。
<br>

**YOLOV5インストール**
Anacondaプロンプトで以下のようにYOLO専用環境を作ってそこで動かす（ほかのパッケージと分けたい）
 conda create -n yolov5 anaconda<br>
 conda activate yolov5 <br>
<br>
git clone https://github.com/ultralytics/yolov5.git <br>
によってyolov5のサブフォルダができて必要なソフトウェアがすべてダウンロードされる<br>
 dir /w<br>
 cd yolov5<br>
 dir /w  
サブフォルダに移動してから以下を実行<br>
pip install -qr requirements.txt<br>
>このチュートリアルの上に同じコマンドあるがノートブックからだと引数が多すぎて実行できない<br>

python detect.py --source 0<br>
でvideo画像に対して物体検知が走るはずである。
>encodingエラーが出たとき、yolov5\data\coco128.yamlをunicodeから
>SJISに変えて保存する。<br>
>ただし、これだとGoogleColabなどで学習したモデルが動かなくなるのでｋの環境構築の前に
>Windowsの環境変数にPYTHONUTF8=1というものを置いて環境をつくる。
>ただしこれまで作ったBASEのPythonコードが動かなかったり、DOS窓のコマンドがうまくでなかったりする＝dir /wが動かなかった

# SJIS環境で独自モデルの実行がうまくいかなかったときのログの例
解決策は現在ベストはPythonutf8=1の環境変数の設定

(yolov5) PS C:\Users\81805\yolov5> python detect.py --source 0 --weights best.pt --conf 0.25 --imgsz 320
detect: weights=['best.pt'], source=0, data=data\coco128.yaml, imgsz=[320, 320], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 v6.1-386-g2e57b84 Python-3.8.13 torch-1.12.1+cpu CPU

Traceback (most recent call last):
  File "detect.py", line 257, in <module>
    main(opt)
  File "detect.py", line 252, in main
    run(**vars(opt))
  File "C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "detect.py", line 93, in run
    model = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)
  File "C:\Users\81805\yolov5\models\common.py", line 334, in __init__
    model = attempt_load(weights if isinstance(weights, list) else w, device=device, inplace=True, fuse=fuse)
  File "C:\Users\81805\yolov5\models\experimental.py", line 80, in attempt_load
    ckpt = torch.load(attempt_download(w), map_location='cpu')  # load
  File "C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\serialization.py", line 712, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\serialization.py", line 1049, in _load
    result = unpickler.load()
  File "C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\serialization.py", line 1042, in find_class
    return super().find_class(mod_name, name)
AttributeError: Can't get attribute 'DetectionModel' on <module 'models.yolo' from 'C:\\Users\\81805\\yolov5\\models\\yolo.py'>

<a align="left" href="https://ultralytics.com/yolov5" target="_blank">
<img width="1024", src="https://user-images.githubusercontent.com/26833433/125273437-35b3fc00-e30d-11eb-9079-46f313325424.png"></a>

これは公式の **YOLOv5 🚀 notebook**です。**Ultralytics**が提供してます。そして自由に配布してもらって構いません。次のライセンスに従います [GPL-3.0 license](https://choosealicense.com/licenses/gpl-3.0/). 
さらなる情報はこちらへ→ https://github.com/ultralytics/yolov5 and https://ultralytics.com. Thank you!

# 設定

レポジトリのクローン、依存性のあるモノのインストール、PyTorchとGPUのチェック

In [1]:
#!git clone https://github.com/ultralytics/yolov5  # clone
#%cd yolov5
#%pip install -qr requirements.txt  # install
#訳注：coco128.yamlをSJISで保存しなおすなどしないとエラーになることがある
#またAnacondaに仮想環境作ってインストールするほうがいい（不要になったらPyTorch毎捨てる・・重たいので
#
#
#

import torch
import utils
display = utils.notebook_init()  # checks

YOLOv5 🚀 v6.1-386-g2e57b84 Python-3.8.13 torch-1.12.1+cpu CPU


Setup complete  (12 CPUs, 31.8 GB RAM, 333.7/1906.5 GB disk)


# 1. 推論

`detect.py` によって様々なソース画像から YOLOv5 の推論を実行できます。モデルは自動的にこちらからダウンロードされます [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases) 、また結果のセーブは `runs/detect`に入ります。
例として推論は以下のようなソースから動かします:

```shell
python detect.py --source 0  # カメラwebcam
                          img.jpg  # 静止画像image 
                          vid.mp4  # ビデオvideo
                          path/  # directory
                          'path/*.jpg'  # glob
                          'https://youtu.be/Zgi9g1ksQHc'  # YouTube
                          'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```

In [1]:
#初期イメージ置くフォルダに静止画像やムービーデータを置く場合（もともとのココモデルの場合：自作モデルの場合は次のコードブロック参照
!python detect.py --source data/images

python: can't open file 'detect.py': [Errno 2] No such file or directory


In [2]:
#Google Colabではリアルタイムはできないのでローカル実行のみ
!python detect.py  --source 0

^C


In [6]:
!python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images
#display.Image(filename='runs/detect/exp/zidane.jpg', width=600)

detect: weights=['yolov5s.pt'], source=data/images, data=data\coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 \U0001f680 v6.1-386-g2e57b84 Python-3.8.13 torch-1.12.1+cpu CPU

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
image 1/5 C:\Users\81805\yolov5\data\images\OIP.jpg: 448x640 8 persons, Done. (0.209s)
image 2/5 C:\Users\81805\yolov5\data\images\S__7086109.jpg: 480x640 4 teddy bears, Done. (0.238s)
image 3/5 C:\Users\81805\yolov5\data\images\WIN_20220812_23_16_37_Pro.jpg: 384x640 1 bed, 2 teddy bears, Done. (0.203s)
image 4/5 C:\Users\81805\yolov5\data\images\bus.jpg: 640x480 4 persons, 1 bus, Done. (0.236s)
image 5/5 C:\User

In [5]:
#自作モデルの実行　YOLOV5の直下にbest.ptを置いて、Webカメラのビデオを置く場合
#Google Colabではリアルタイムはできないのでローカル実行のみ
!python detect.py --source 0 --weights best.pt --conf 0.25

detect: weights=['best.pt'], source=0, data=data\coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 \U0001f680 v6.1-386-g2e57b84 Python-3.8.13 torch-1.12.1+cpu CPU

Traceback (most recent call last):
  File "detect.py", line 257, in <module>
    main(opt)
  File "detect.py", line 252, in main
    run(**vars(opt))
  File "C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "detect.py", line 93, in run
    model = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)
  File "C:\Users\81805\yolov5\models\common.py", line 334, in __init__
    model 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img align="left" src="https://user-images.githubusercontent.com/26833433/127574988-6a558aa1-d268-44b9-bf6b-62d4c605cc72.jpg" width="600">

# 2. 検証
Validate a model's accuracy on [COCO](https://cocodataset.org/#home) val or test-dev datasets. Models are downloaded automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases). To show results by class use the `--verbose` flag. Note that `pycocotools` metrics may be ~1% better than the equivalent repo metrics, as is visible below, due to slight differences in mAP computation.

## COCO val
Download [COCO val 2017](https://github.com/ultralytics/yolov5/blob/74b34872fdf41941cddcf243951cdb090fbac17b/data/coco.yaml#L14) dataset (1GB - 5000 images), and test model accuracy.

In [ ]:
# Download COCO val
torch.hub.download_url_to_file('https://ultralytics.com/assets/coco2017val.zip', 'tmp.zip')
!unzip -q tmp.zip -d ../datasets && rm tmp.zip

  0%|          | 0.00/780M [00:00<?, ?B/s]

In [ ]:
# Run YOLOv5x on COCO val
!python val.py --weights yolov5x.pt --data coco.yaml --img 640 --iou 0.65 --half

val: data=/content/yolov5/data/coco.yaml, weights=['yolov5x.pt'], batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.65, task=val, device=, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs/val, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 v6.1-370-g20f1b7e Python-3.7.13 torch-1.12.0+cu113 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)

100% 166M/166M [00:35<00:00, 4.97MB/s]

Fusing layers... 
YOLOv5x summary: 444 layers, 86705005 parameters, 0 gradients
100% 755k/755k [00:00<00:00, 49.4MB/s]
val: Scanning '/content/datasets/coco/val2017' images and labels...4952 found, 48 missing, 0 empty, 0 corrupt: 100% 5000/5000 [00:00<00:00, 10716.86it/s]
val: New cache created: /content/datasets/coco/val2017.cache
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100% 157/157 [01:08<00:00,  2.28it/s]
                 all       5000      36335      0.743      0.6

## COCO test
Download [COCO test2017](https://github.com/ultralytics/yolov5/blob/74b34872fdf41941cddcf243951cdb090fbac17b/data/coco.yaml#L15) dataset (7GB - 40,000 images), to test model accuracy on test-dev set (**20,000 images, no labels**). Results are saved to a `*.json` file which should be **zipped** and submitted to the evaluation server at https://competitions.codalab.org/competitions/20794.

In [ ]:
# Download COCO test-dev2017
torch.hub.download_url_to_file('https://ultralytics.com/assets/coco2017labels.zip', 'tmp.zip')
!unzip -q tmp.zip -d ../datasets && rm tmp.zip
!f="test2017.zip" && curl http://images.cocodataset.org/zips/$f -o $f && unzip -q $f -d ../datasets/coco/images

In [ ]:
# Run YOLOv5x on COCO test
!python val.py --weights yolov5x.pt --data coco.yaml --img 640 --iou 0.65 --half --task test

# 以下のRoboflow今のところうまく動いてない
（オンラインはよく動いた）ダウンロードしたデータセットのアノテーション座標が悪い。
<br>
labelimgという古くからあるアノテーションツールで実行する<br>
https://github.com/tzutalin/labelImg/releases
<br>
使い方の参考サイト
https://laid-back-scientist.com/labelimg

# 3. 学習

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="1000" src="https://uploads-ssl.webflow.com/5f6bc60e665f54545a1e52a5/615627e5824c9c6195abfda9_computer-vision-cycle.png"/></a></p>
 `roboflow` pip パッケージを使って独自の推論条件から画像をサンプリングして、アクティブな学習ループを構築しましょう
<br><br>

YOLOv5s モデルを [COCO128](https://www.kaggle.com/ultralytics/coco128) datasetで `--data coco128.yaml`, で事前学習された `--weights yolov5s.pt`モデルから、あるいはランダムで初期化された `--weights '' --cfg yolov5s.yaml`.
から学習させてみましょう。
- **Pretrained [Models](https://github.com/ultralytics/yolov5/tree/master/models)** は次から自動的にダウンロードされます [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases)
- **[Datasets](https://github.com/ultralytics/yolov5/tree/master/data)** は自動ダウンロードされ以下のものを含みます: [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [Argoverse](https://github.com/ultralytics/yolov5/blob/master/data/Argoverse.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml), [xView](https://github.com/ultralytics/yolov5/blob/master/data/xView.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [SKU-110K](https://github.com/ultralytics/yolov5/blob/master/data/SKU-110K.yaml).
- **学習結果** 次にセーブされます `runs/train/` これは毎回番号があがるディレクトリで例えば次のように増えていきます `runs/train/exp2`, `runs/train/exp3` etc.
<br><br>

**Mosaic Dataloader** は４つの画像を一つに組み合わせるような学習に使います

## Roboflowを利用したカスタムデータの学習 🌟 NEW

[Roboflow](https://roboflow.com/?ref=ultralytics) はユーザーの独自データから簡単に高品質なデータセットを **集めて、ラベル付けし、準備する** ことを可能にします。 Roboflowはまたアクティブなラーニングのパイプラインを簡単に構築したり、チームでデータセットを改善することを可能にします。そして直接ユーザーのモデル構築ワークフローに`roboflow`PIPパッケージとして統合されています。

- カスタム学習の例: [https://blog.roboflow.com/how-to-train-yolov5-on-a-custom-dataset/](https://blog.roboflow.com/how-to-train-yolov5-on-a-custom-dataset/?ref=ultralytics)
- Custom Training Notebook: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/roboflow-ai/yolov5-custom-training-tutorial/blob/main/yolov5-custom-training.ipynb)
<br>

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="480" src="https://uploads-ssl.webflow.com/5f6bc60e665f54545a1e52a5/6152a275ad4b4ac20cd2e21a_roboflow-annotate.gif"/></a></p>非常に高速なラベリングができます (モデル支援型のラべリングを含みます)

In [ ]:
# Tensorboard  (optional)
%load_ext tensorboard
%tensorboard --logdir runs/train

In [ ]:
# ClearML  (optional)
%pip install -q clearml
!clearml-init

In [ ]:
# Weights & Biases  (optional)
%pip install -q wandb
import wandb
wandb.login()

In [1]:
# Train YOLOv5s on COCO128 for 3 epochs
!python train.py --img 416 --batch 20 --epochs 3 --data ./data.yaml --weights yolov5s.pt --cache

In [ ]:
#GCPで使ったスクリプトdatasetsの直下にtrainとvalid（それぞれの下にimagesとlabelsおいた）
!python train.py --img 416 --batch 16 --epochs 300 --patience 0 --data data_gcp.yaml --weights yolov5s.pt --cache

# 4. Visualize

## ClearML Logging and Automation 🌟 NEW

[ClearML](https://cutt.ly/yolov5-notebook-clearml) is completely integrated into YOLOv5 to track your experimentation, manage dataset versions and even remotely execute training runs. To enable ClearML (check cells above):

- `pip install clearml`
- run `clearml-init` to connect to a ClearML server (**deploy your own [open-source server](https://github.com/allegroai/clearml-server)**, or use our [free hosted server](https://cutt.ly/yolov5-notebook-clearml))

You'll get all the great expected features from an experiment manager: live updates, model upload, experiment comparison etc. but ClearML also tracks uncommitted changes and installed packages for example. Thanks to that ClearML Tasks (which is what we call experiments) are also reproducible on different machines! With only 1 extra line, we can schedule a YOLOv5 training task on a queue to be executed by any number of ClearML Agents (workers).

You can use ClearML Data to version your dataset and then pass it to YOLOv5 simply using its unique ID. This will help you keep track of your data without adding extra hassle. Explore the [ClearML Tutorial](https://github.com/ultralytics/yolov5/tree/master/utils/loggers/clearml) for details!

<a href="https://cutt.ly/yolov5-notebook-clearml">
<img alt="ClearML Experiment Management UI" src="https://github.com/thepycoder/clearml_screenshots/raw/main/scalars.jpg" width="1280"/></a>

## Weights & Biases Logging

[Weights & Biases](https://wandb.ai/site?utm_campaign=repo_yolo_notebook) (W&B) is integrated with YOLOv5 for real-time visualization and cloud logging of training runs. This allows for better run comparison and introspection, as well improved visibility and collaboration for teams. To enable W&B `pip install wandb`, and then train normally (you will be guided through setup on first use). 

During training you will see live updates at [https://wandb.ai/home](https://wandb.ai/home?utm_campaign=repo_yolo_notebook), and you can create and share detailed [Reports](https://wandb.ai/glenn-jocher/yolov5_tutorial/reports/YOLOv5-COCO128-Tutorial-Results--VmlldzozMDI5OTY) of your results. For more information see the [YOLOv5 Weights & Biases Tutorial](https://github.com/ultralytics/yolov5/issues/1289).  

<a href="https://wandb.ai/glenn-jocher/yolov5_tutorial">
<img alt="Weights & Biases dashboard" src="https://user-images.githubusercontent.com/26833433/182482859-288a9622-4661-48db-99de-650d1dead5c6.jpg" width="1280"/></a>

## Local Logging

Training results are automatically logged with [Tensorboard](https://www.tensorflow.org/tensorboard) and [CSV](https://github.com/ultralytics/yolov5/pull/4148) loggers to `runs/train`, with a new experiment directory created for each new training as `runs/train/exp2`, `runs/train/exp3`, etc.

This directory contains train and val statistics, mosaics, labels, predictions and augmentated mosaics, as well as metrics and charts including precision-recall (PR) curves and confusion matrices. 

<img alt="Local logging results" src="https://user-images.githubusercontent.com/26833433/183222430-e1abd1b7-782c-4cde-b04d-ad52926bf818.jpg" width="1280"/>


# Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


# Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on macOS, Windows, and Ubuntu every 24 hours and on every commit.


# Appendix

Optional extras below. Unit tests validate repo functionality and should be run on any PRs submitted.


In [ ]:
# Reproduce
for x in 'yolov5n', 'yolov5s', 'yolov5m', 'yolov5l', 'yolov5x':
  !python val.py --weights {x}.pt --data coco.yaml --img 640 --task speed  # speed
  !python val.py --weights {x}.pt --data coco.yaml --img 640 --conf 0.001 --iou 0.65  # mAP

In [ ]:
# PyTorch Hub
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()

In [ ]:
# YOLOv5 CI
%%shell
rm -rf runs  # remove runs/
m=yolov5n  # official weights
b=runs/train/exp/weights/best  # best.pt checkpoint
python train.py --imgsz 64 --batch 32 --weights $m.pt --cfg $m.yaml --epochs 1 --device 0  # train
for d in 0 cpu; do  # devices
  for w in $m $b; do  # weights
    python val.py --imgsz 64 --batch 32 --weights $w.pt --device $d  # val
    python detect.py --imgsz 64 --weights $w.pt --device $d  # detect
  done
done
python hubconf.py --model $m  # hub
python models/tf.py --weights $m.pt  # build TF model
python models/yolo.py --cfg $m.yaml  # build PyTorch model
python export.py --weights $m.pt --img 64 --include torchscript  # export

In [ ]:
# Profile
from utils.torch_utils import profile

m1 = lambda x: x * torch.sigmoid(x)
m2 = torch.nn.SiLU()
results = profile(input=torch.randn(16, 3, 640, 640), ops=[m1, m2], n=100)

In [ ]:
# Evolve
!python train.py --img 640 --batch 64 --epochs 100 --data coco128.yaml --weights yolov5s.pt --cache --noautoanchor --evolve
!d=runs/train/evolve && cp evolve.* $d && zip -r evolve.zip $d && gsutil mv evolve.zip gs://bucket  # upload results (optional)

In [ ]:
# VOC
for b, m in zip([64, 64, 64, 32, 16], ['yolov5n', 'yolov5s', 'yolov5m', 'yolov5l', 'yolov5x']):  # batch, model
  !python train.py --batch {b} --weights {m}.pt --data VOC.yaml --epochs 50 --img 512 --hyp hyp.VOC.yaml --project VOC --name {m} --cache

In [ ]:
# TensorRT 
!pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com  # install
!python export.py --weights yolov5s.pt --include engine --imgsz 640 --device 0  # export
!python detect.py --weights yolov5s.engine --imgsz 640 --device 0  # inference

In [18]:
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)

# Images
imgs = ['https://ultralytics.com/images/zidane.jpg']  # batch of images

# Inference
results = model(imgs)

# Results
results.print()
results.save()  # or .show()

results.xyxy[0]  # img1 predictions (tensor)
results.pandas().xyxy[0]  # img1 predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 1  433.50  433.50   517.5  714.5    0.687988     27     tie
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

C:\Users\81805\anaconda3\envs\yolov5\lib\site-packages\torch\hub.py:266: UserWarning: You are about to download and run code from an untrusted repository. In a future release, this won't be allowed. To add the repository to your trusted list, change the command to {calling_fn}(..., trust_repo=False) and a command prompt will appear asking for an explicit confirmation of trust, or load(..., trust_repo=True), which will assume that the prompt is to be answered with 'yes'. You can also use load(..., trust_repo='check') which will only prompt for confirmation if the repo is not already trusted. This will eventually be the default behaviour
  warnings.warn(
Downloading: "https://github.com/ultralytics/yolov5/zipball/master" to C:\Users\81805/.cache\torch\hub\master.zip
YOLOv5  2022-8-17 Python-3.8.13 torch-1.12.1+cpu CPU

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients
Adding AutoShape... 
Saved 1 image to runs\detect\exp49


image 1/1: 720x1280 2 persons, 2 ties
Speed: 200.9ms pre-process, 124.5ms inference, 3.0ms NMS per image at shape (1, 3, 384, 640)


xmin        ymin         xmax        ymax  confidence  class    name
0  742.974792   48.395599  1141.844604  720.000000    0.881052      0  person
1  442.007660  437.522461   496.653961  709.973511    0.675214     27     tie
2  123.024261  193.287292   715.662109  719.723877    0.665813      0  person
3  982.803162  308.417358  1027.365845  419.987000    0.260076     27     tie